app/models/proposal.rb

```diff
@@ -118,13 +124,15 @@ def withdraw
   end
 
   def confirm
-    transaction do
-      update!(confirmed_at: DateTime.current)
-      ps = ProgramSession.create_from_proposal(self)
-      ps.persisted?
-    end
-  rescue ActiveRecord::RecordInvalid
-    false
+    self.update(confirmed_at: DateTime.current)
+    ps = self.program_session
+    ps.update(state: self.waitlisted? ? ProgramSession::WAITLISTED : ProgramSession::LIVE) if ps.present?
+  end
+
+  def decline
+    self.update(state: WITHDRAWN)
```

Comment:
Look at removing self.  Then combine the two updates into a single call.  Right now it's doing two database hits for updating the proposal.  update can take multiple attributes so there's no reason not to combine them.